builtin/v11/migration/miner.go

```diff
@@ -0,0 +1,323 @@
+
+func (m minerMigrator) migrateDeadlines(ctx context.Context, store adt10.Store, cache migration.MigrationCache, deadlines cid.Cid) (cid.Cid, error) {
+	if deadlines == m.emptyDeadlinesV10 {
+		return m.emptyDeadlinesV11, nil
+	}
+
+	var inDeadlines miner10.Deadlines
+	err := store.Get(store.Context(), deadlines, &inDeadlines)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	var outDeadlines miner11.Deadlines
+	for i, c := range inDeadlines.Due {
+		if c == m.emptyDeadlineV10 {
+			outDeadlines.Due[i] = m.emptyDeadlineV11
+		} else {
+			var inDeadline miner10.Deadline
+			if err = store.Get(ctx, c, &inDeadline); err != nil {
+				return cid.Undef, err
+			}
+
+			outSectorsSnapshotCid, err := cache.Load(migration.SectorsAmtKey(inDeadline.SectorsSnapshot), func() (cid.Cid, error) {
+				inSectorsSnapshot, err := adt10.AsArray(store, inDeadline.SectorsSnapshot, miner10.SectorsAmtBitwidth)
+				if err != nil {
+					return cid.Undef, err
+				}
+
+				outSectorsSnapshot, err := migrateSectorsFromScratch(ctx, store, inSectorsSnapshot)
+				if err != nil {
+					return cid.Undef, xerrors.Errorf("failed to migrate sectors: %w", err)
+				}
+
+				return outSectorsSnapshot.Root()
+			})
+
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to migrate sectors snapshot: %w", err)
+			}
+
+			outDeadline := miner11.Deadline{
+				Partitions:                        inDeadline.Partitions,
+				ExpirationsEpochs:                 inDeadline.ExpirationsEpochs,
+				PartitionsPoSted:                  inDeadline.PartitionsPoSted,
+				EarlyTerminations:                 inDeadline.EarlyTerminations,
+				LiveSectors:                       inDeadline.LiveSectors,
+				TotalSectors:                      inDeadline.TotalSectors,
+				FaultyPower:                       miner11.PowerPair(inDeadline.FaultyPower),
+				OptimisticPoStSubmissions:         inDeadline.OptimisticPoStSubmissions,
+				SectorsSnapshot:                   outSectorsSnapshotCid,
+				PartitionsSnapshot:                inDeadline.PartitionsSnapshot,
+				OptimisticPoStSubmissionsSnapshot: inDeadline.OptimisticPoStSubmissionsSnapshot,
+			}
+
+			outDlCid, err := store.Put(ctx, &outDeadline)
+			if err != nil {
+				return cid.Undef, err
+			}
+
+			outDeadlines.Due[i] = outDlCid
+		}
+	}
+
+	return store.Put(ctx, &outDeadlines)
+}
+
+func migrateSectorInfo(sectorInfo miner10.SectorOnChainInfo) *miner11.SectorOnChainInfo {
+	// For a sector that has not been updated: the Activation is correct and ReplacedSectorAge is zero.
+	// For a sector that has been updated: Activation is the epoch at which it was upgraded, and ReplacedSectorAge is delta since the true activation.
+	//
+	// Thus we want to set:
+	//
+	// Activation := OldActivation - ReplacedSectorAge (a no-op for non-upgraded sectors)
+	// PowerBaseEpoch := Activation (in both upgrade and not-upgraded cases)
+
+	powerBaseEpoch := sectorInfo.Activation
+	activationEpoch := sectorInfo.Activation - sectorInfo.ReplacedSectorAge
```

Comment:
Curious here -- do we have any really old school CC upgrade sectors with a non zero Replaced sector age but a true activation time set at Activation ?  Its worth raising a flag in a practice migration to see if this is something we need to handle.  I think we are safe if we only make this change if SectorKey != nil